bn-apps/ledger-sync/ledger-sync-service/src/main/kotlin/net/corda/businessnetworks/ledgersync/TransactionRecoveryFlow.kt

```diff
@@ -0,0 +1,58 @@
+package net.corda.businessnetworks.ledgersync
+
+import co.paralleluniverse.fibers.Suspendable
+import net.corda.core.crypto.SecureHash
+import net.corda.core.flows.FlowLogic
+import net.corda.core.flows.FlowSession
+import net.corda.core.flows.InitiatedBy
+import net.corda.core.flows.InitiatingFlow
+import net.corda.core.flows.ReceiveTransactionFlow
+import net.corda.core.flows.SendTransactionFlow
+import net.corda.core.identity.Party
+import net.corda.core.utilities.unwrap
+
+/**
+ * This flow allows for the recovery of transactions with IDs found previously.
+ */
+@InitiatingFlow
+class TransactionRecoveryFlow(
+        private val report: Map<Party, LedgerSyncFindings>
+) : FlowLogic<Unit>() {
+
+    @Suspendable
+    override fun call() {
+        report.toList().map { (counterParty, findings) ->
+            val session = initiateFlow(counterParty)
+            findings.recoverable().also {
+                session.send(it)
```

Comment:
Might make sense to double check whether a tx exists in the vault, before syncing it. Trilateral (and more-lateral) transactions might be already recovered as a part of syncs with previous members. Or alternatevely we can filter out all duplicate transaction ids even before starting a sync.

Reply:
Using `serviceHub.validatedTransactions.getTransaction()` to validate finds a transaction regardless (potentially by using a cache).
`serviceHub.vaultService.queryBy(ContractState::class.java, VaultQueryCriteria(stateRefs = ...))` requires me to pass in an `index` as well, which I don't have available.